### PR TITLE
fat_fs: Add support for the "win" variable alignment in FATFS struct

### PR DESCRIPTION
**Pull request by ngboonkhai**
This is to support Sypnosys Designware SDMMC controller read and write
operation where the buffer address is needed to be 16bytes and 512bytes
aligned. Adding macro FS_FATFS_WIN_ALIGN to align the "win" variable
address in FATFS struct.

Signed-off-by: Boon Khai Ng <boon.khai.ng@intel.com>

### CODE REVIEW

**Comment by ngboonkhai** [comment]
Hi all, 
I tried to find the add reviewer button but no luck :( .
Refer to previous PR, I saw your name reviewing the ffconf.h, hence adding you all and need your help to review this PR
@pabigot 
@de-nordic
@carlescufi
@nashif 

Regards,
Boon Khai

**Comment by ngboonkhai** [comment]
Hi @de-nordic  @carlescufi ,

I'm still pending review for this PR, can you please help to review?

Regards,
Boon Khai

**Comment by ngboonkhai** [comment]
Hi @de-nordic  @carlescufi 

Any updates?

Regards, 
Boon Khai

File: include/ff.h
Comment (de-nordic) [inline include/ff.h:176]: Wouldn't it work to always default `FS_FATFS_WIN_ALIGNED` to 1 when not set otherwise and remove the ifdef completely?

Comment (ngboonkhai) [reply include/ff.h:176]: Hi Ermel,

I was using the FS_FATFS_WIN_ALIGNED as the alignment value as well. See line 177.

In Zephyr, there is a config setting to define CONFIG_FS_FATFS_WIN_ALIGNED=512, meaning to align the win variable with 512 bytes, if setting the CONFIG_FS_FATFS_WIN_ALIGNED=16, it will align the win variable with 16 bytes.

Regards, 
Boon Khai

Comment (de-nordic) [reply include/ff.h:176]: what I mean is that if you do something like:
```
#if defined(CONFIG_FS_FATFS_WIN_ALIGNED)
#define FS_FATFS_WIN_ALIGNED CONFIG_FS_FATFS_WIN_ALIGNED
#else
#define FS_FATFS_WIN_ALIGNED 1
#fi
```
then is the `#ifdef still needed`?
or can it be just `BYTE win[FF_MAX_SS] __attribute__((aligned(FS_FATFS_WIN_ALIGNED)))`
where default value of 1 for `FS_FATFS_WIN_ALIGNED` would mean "no alignment".

Also, this seems problematic because if you set the alignment to 512 you may end up with a lot of wasted space in the structure.

Comment (ngboonkhai) [reply include/ff.h:176]: Hi Ermel, what I did is, (refer to ffconf.h)

#if defined(CONFIG_FS_FATFS_WIN_ALIGNED)
#if CONFIG_FS_FATFS_WIN_ALIGNED != 0
#define FS_FATFS_WIN_ALIGNED   CONFIG_FS_FATFS_WIN_ALIGNED
#endif /* CONFIG_FS_FATFS_WIN_ALIGNED != 0 */
#endif 

If the setting CONFIG_FS_FATFS_WIN_ALIGNED was not set, it will be 0 by default. When it is zero, then FS_FATFS_WIN_ALIGNED will not defined. When FS_FATFS_WIN_ALIGNED is not defined, the original definition of BYTE win[FF_MAX_SS] will be used without alignment. 

From our side, the SDMMC driver is copied block by block and need to be 512 byte aligned, and that's why I implement this CONFIG for user to choose whether or not to align the WIN. For those who don't have the hard requirement, they will not enable this CONFIG thus saving the space in the structure.

**Comment by carlescufi** [comment]
@ngboonkhai we don't really have any additional mantainers besides @de-nordic. As soon as he approves we can merge this, but remember that you need to create a PR to the main zephyr repo as described here:
https://docs.zephyrproject.org/latest/guides/modules.html#process-for-submitting-changes-to-existing-modules

File: include/ffconf.h
Comment (de-nordic) [inline include/ffconf.h:260]: Do the `#ifdef` vlock similar way the other Kconfig configured ff parameters are, look for example how ``CONFIG_FS_FATFS_MAX_SS` affects `FF_MAX_SS`.

Comment (de-nordic) [reply include/ffconf.h:260]: And change the name from `WIN_ALIGNED` to `WINDOW_ALIGNMENT`, because we do not anything to `WIN` we aligned `WIN`  to something.

File: include/ff.h
Comment (de-nordic) [reply include/ff.h:176]: Can you check how does the `__attribute__((aligned(1)));` behaves, because if it makes default alignment of byte array then the `#ifdef FS_FATFS_WIN_ALIGNED` s not needed: it would be enough, in such case, to do `BYTE win[FF_MAX_SS] __attribute__((aligned(FS_FATFS_WIN_ALIGNED)))` with no ifdef, because default value of `FS_FATFS_WIN_ALGNED=1` would work fine.

File: include/ffconf.h
Comment (de-nordic) [reply include/ffconf.h:260]: Referring to https://github.com/zephyrproject-rtos/fatfs/pull/9/files#r774606396 if `CONFIG_FS_FATFS_WINDOW_ALIGNMENT` is not defined, default the FS_FATFS_WINDOW_ALIGNMENT to 1

File: include/ff.h
Comment (ngboonkhai) [reply include/ff.h:176]: > Can you check how does the `__attribute__((aligned(1)));` behaves, because if it makes default alignment of byte array then the `#ifdef FS_FATFS_WIN_ALIGNED` s not needed: it would be enough, in such case, to do `BYTE win[FF_MAX_SS] __attribute__((aligned(FS_FATFS_WIN_ALIGNED)))` with no ifdef, because default value of `FS_FATFS_WIN_ALGNED=1` would work fine.

Hi Ermel,

My bad, `__attribute__((aligned(1)));` actually works.
I agree with you suggestion to set default the value with 1, and with the following changes in the ffconf.h
```
#if defined(CONFIG_FS_FATFS_WIN_ALIGNED)
#define FS_FATFS_WIN_ALIGNED CONFIG_FS_FATFS_WIN_ALIGNED
#else
#define FS_FATFS_WIN_ALIGNED 1
#endif
```

**Comment by ngboonkhai** [comment]
> @ngboonkhai we don't really have any additional mantainers besides @de-nordic. As soon as he approves we can merge this, but remember that you need to create a PR to the main zephyr repo as described here: https://docs.zephyrproject.org/latest/guides/modules.html#process-for-submitting-changes-to-existing-modules

Okay noted carlescufi. Will work closely with Ermel.

File: include/ffconf.h
Comment (ngboonkhai) [reply include/ffconf.h:260]: Okay, will update the PR again with the above suggestion. 
1) To use follow the #ifdef like what FF_MAX_SS defined
2) change the WIN_ALIGNED to WINDOW_ALIGNMENT
3) default the FS_FATFS_WINDOW_ALIGNMENT to 1

**Comment by ngboonkhai** [comment]
Hi Ermel,

Done force push the PR with the update that align with our discussion. Please check.

Regards,
Boon Khai.

**Review by de-nordic — APPROVED** [review]
Looks OK. Thanks!

**Comment by ngboonkhai** [comment]
> Looks OK. Thanks!

Thanks @de-nordic :D

**Comment by ngboonkhai** [comment]
@carlescufi 
Hi Carles,

Please help me to merge the PR, Thanks 

Regards,
Boon Khai

**Comment by de-nordic** [comment]
> @carlescufi Hi Carles,
> 
> Please help me to merge the PR, Thanks
> 
> Regards, Boon Khai

@ngboonkhai Please address my issue here https://github.com/zephyrproject-rtos/zephyr/pull/41096#discussion_r776768381

**Comment by ngboonkhai** [comment]
> > @carlescufi Hi Carles,
> > Please help me to merge the PR, Thanks
> > Regards, Boon Khai
> 
> @ngboonkhai Please address my issue here [zephyrproject-rtos/zephyr#41096 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/41096#discussion_r776768381)

Hi @de-nordic 

Done force push, Thanks

Regards,
Boon Khai

**Comment by de-nordic** [comment]
@carlescufi This already passes tests with Zephyr manifest, so please review.

**Comment by de-nordic** [comment]
@carles @galak @nashif, anyone please review. We need second review for this.

**Review by nvlsianpu — APPROVED** [review]

**Comment by ngboonkhai** [comment]
Thanks @de-nordic, @nvlsianpu !

**Pull request merged**
